### ISSUE TITLE
DDP Wrapper for L2l datasets

### ISSUE THREAD

**Issue by arijitthegame** [learnables/learn2learn#263]
Hi, 

Thank you for the excellent repo. I am wondering if there is any DDP wrapper for the L2L datasets. My main use case is protonets. I saw this post https://github.com/learnables/learn2learn/issues/170 and I am wondering if there is any other solution to MAML or gradient based methods with just DDP and not pytorch lightning? 

 Thank you.

**Comment by seba-1511** [comment]
Hi @arijitthegame,

Apologies for the delayed answer. At the moment the simplest way to distribute MAML across Python processes is through cherry' `Distributed` optimizer, or to use lightning.

**Comment by brando90** [comment]
> Hi,
> 
> Thank you for the excellent repo. I am wondering if there is any DDP wrapper for the L2L datasets. My main use case is protonets. I saw this post #170 and I am wondering if there is any other solution to MAML or gradient based methods with just DDP and not pytorch lightning?
> 
> Thank you.

why does proro nets care if its lightning or not?

**Comment by seba-1511** [comment]
Hi @arijitthegame and @brando90,

I've just added an example of using the cherry's Distributed optimizer to train MAML on multiple GPUs with `torch.distributed`. Please check: https://github.com/learnables/learn2learn/blob/master/examples/vision/distributed_maml.py

**Comment by brando90** [comment]
@seba-1511 you are amazing seba! Thank you!

**Comment by brando90** [comment]
@seba-1511 would just commenting out the `opt.step()` work for distributed meta-validation evaluation? (I am also not only training of course)

Thanks again1

**Comment by arijitthegame** [comment]
Thank you so much!

**Comment by brando90** [comment]
Seba is too kind but don't forget to cite l2l as gratitude if you use his kindly example of cherry! https://github.com/learnables/learn2learn#citation Thanks again!

**Comment by brando90** [comment]
seba said:

Hi

I've just added an example of using the cherry's Distributed optimizer to train MAML on multiple GPUs with torch.distributed. Please check: https://github.com/learnables/learn2learn/blob/master/examples/vision/distributed_maml.py

**Comment by brando90** [comment]
> Hi @arijitthegame and @brando90,
> 
> I've just added an example of using the cherry's Distributed optimizer to train MAML on multiple GPUs with `torch.distributed`. Please check: https://github.com/learnables/learn2learn/blob/master/examples/vision/distributed_maml.py

curious, what are the pros and cons vs the distributed meta-learning in pytorch lightning?

**Comment by arijitthegame** [comment]
@brando90 Thank you so much for the example. This is merely my personal taste. I love low level APIs. I generally use Apex (cloning from Nvidia) and writing scripts using torch distributed and their launcher. I have always found it useful to write things this way as it forced me to learn how to use these tools and what goes on around the code (not always I understand them but I strive to). The lightning module is excellent as you can compress about 50 lines of code in just a trainer.fit() method with the kwargs. 

The reason I was hesitant to use the lightning for L2L datasets is because I could not figure out how the batches are split across the nodes/gpus. And that's the reason I could not use the usual DDP modules. In some way I still would love to learn is how lightning handles the gradient hooks (in cloning the model) and the splitting across various machines. 

To summarize: this is a personal preference to use low level APIs.

**Comment by brando90** [comment]
> @brando90 Thank you so much for the example. This is merely my personal taste. I love low level APIs. I generally use Apex (cloning from Nvidia) and writing scripts using torch distributed and their launcher. I have always found it useful to write things this way as it forced me to learn how to use these tools and what goes on around the code (not always I understand them but I strive to). The lightning module is excellent as you can compress about 50 lines of code in just a trainer.fit() method with the kwargs.
> 
> The reason I was hesitant to use the lightning for L2L datasets is because I could not figure out how the batches are split across the nodes/gpus. And that's the reason I could not use the usual DDP modules. In some way I still would love to learn is how lightning handles the gradient hooks (in cloning the model) and the splitting across various machines.
> 
> To summarize: this is a personal preference to use low level APIs.

thanks for sharing that!

I am curious, what did you end up using for distributed MAML?

**Comment by brando90** [comment]
@arijitthegame when you mentioned lightining, did you mean this: https://devblog.pytorchlightning.ai/lightning-flash-now-supports-meta-learning-7c0ac8b1cde7 ???

I am curious to know if lightning supports distributed training. Though at this point I decided to go for seba's cherry example + code.

**Comment by brando90** [comment]
@seba-1511 I think there might be a bug in your distributed MAML example. I am still going through it to understand it better but the input to main takes `word_size` but it is not being used inside the main script:

```
def main(
        ways=5,
        shots=5,
        meta_lr=0.003,
        fast_lr=0.5,
        meta_batch_size=32,
        adaptation_steps=1,
        num_iterations=60_000,
        cuda=True,
        rank=0,
        world_size=1,
        seed=42,
):
    print(rank)
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    # todo: Q, could get device be all I need here or do I use the rank?
    device = torch.device('cpu')
    if cuda and torch.cuda.device_count():
        torch.cuda.manual_seed(seed)
        device_id = rank % torch.cuda.device_count()
        device = torch.device('cuda:' + str(device_id))
    print(rank, ':', device)

    # Create Tasksets using the benchmark interface
    tasksets = l2l.vision.benchmarks.get_tasksets(
        'mini-imagenet',
        train_samples=2*shots,
        train_ways=ways,
        test_samples=2*shots,
        test_ways=ways,
        root='~/data',
    )

    # Create model
    model = l2l.vision.models.MiniImagenetCNN(ways)
    model.to(device)
    maml = l2l.algorithms.MAML(model, lr=fast_lr, first_order=False)
    opt = torch.optim.Adam(maml.parameters(), meta_lr)
    opt = cherry.optim.Distributed(maml.parameters(), opt=opt, sync=1)
    opt.sync_parameters()
    loss = torch.nn.CrossEntropyLoss(reduction='mean')
    print(rank, ':', device)

    for iteration in range(num_iterations):
        opt.zero_grad()
        meta_train_error = 0.0
        meta_train_accuracy = 0.0
        meta_valid_error = 0.0
        meta_valid_accuracy = 0.0
        for task in range(meta_batch_size):
            # Compute meta-training loss
            learner = maml.clone()
            batch = tasksets.train.sample()
            evaluation_error, evaluation_accuracy = fast_adapt(
                batch,
                learner,
                loss,
                adaptation_steps,
                shots,
                ways,
                device,
            )
            evaluation_error.backward()
            meta_train_error += evaluation_error.item()
            meta_train_accuracy += evaluation_accuracy.item()

            # Compute meta-validation loss
            learner = maml.clone()
            batch = tasksets.validation.sample()
            evaluation_error, evaluation_accuracy = fast_adapt(
                batch,
                learner,
                loss,
                adaptation_steps,
                shots,
                ways,
                device,
            )
            meta_valid_error += evaluation_error.item()
            meta_valid_accuracy += evaluation_accuracy.item()

        # Print some metrics
        if rank == 0:
            print('\n')
            print('Iteration', iteration)
            print('Meta Train Error', meta_train_error / meta_batch_size)
            print('Meta Train Accuracy', meta_train_accuracy / meta_batch_size)
            print('Meta Valid Error', meta_valid_error / meta_batch_size)
            print('Meta Valid Accuracy', meta_valid_accuracy / meta_batch_size)

        # Average the accumulated gradients and optimize
        for p in maml.parameters():
            p.grad.data.mul_(1.0 / meta_batch_size)
        opt.step()  # averages gradients across all workers

    meta_test_error = 0.0
    meta_test_accuracy = 0.0
    for task in range(meta_batch_size):
        # Compute meta-testing loss
        learner = maml.clone()
        batch = tasksets.test.sample()
        evaluation_error, evaluation_accuracy = fast_adapt(
            batch,
            learner,
            loss,
            adaptation_steps,
            shots,
            ways,
            device,
        )
        meta_test_error += evaluation_error.item()
        meta_test_accuracy += evaluation_accuracy.item()
    print('Meta Test Error', meta_test_error / meta_batch_size)
    print('Meta Test Accuracy', meta_test_accuracy / meta_batch_size)
 ```

what was `world_size` meant to be doing?

**Comment by brando90** [comment]
is the `-m` in `    python -m torch.distributed.launch --nproc_per_node=2 distributed_maml.py` really needed? `-m` runs a module as a script but we are passing a script as input so idk if that is needed.

**Comment by brando90** [comment]
@seba-1511 I noticed that you have this 

`        init_method='file:///home/seba-1511/.dist_init`

can you provide what that file has please? 

Thanks again for your help! :)

**Comment by brando90** [comment]
also, fyi, seems that `torch.distributed.launch` will be deprecated in favour of `torchrun` see: https://pytorch.org/docs/stable/elastic/run.html#launcher-api and https://pytorch.org/docs/stable/distributed.html

**Comment by brando90** [comment]
I'm getting this warning:
```
WARNING:torch.distributed.run:*****************************************
Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed.
```
does someone know how to select that env var? is the default 1 a good value? if yes why?

Thanks!

**Comment by brando90** [comment]
I'm using a cosine scheduler. Does that interact in any way with cherry? My supposition is that each process "has a copy of the running script" and update their local parameters and scheduler on their own. Is that a correct assumption? @seba-1511 

Apologies for the pings...your time is really appreciated. Thanks in advance!

**Comment by brando90** [comment]
curious, in DDP there is usually a `cleanup()` code that is run at the end of the entire script. But I noticed that your code did not have that. Is that a small bug? In ddp I usually run:

```
def cleanup(rank):
    """ Destroy a given process group, and deinitialize the distributed package """
    # only destroy the process distributed group if the code is not running serially
    if is_running_parallel(rank):
        torch.distributed.barrier()
        dist.destroy_process_group()
```
is that not needed anymore?

**Comment by brando90** [comment]
is it a problem to give the params to the optimizer like this isntead?

```
    opt = torch.optim.Adam(model.parameters(), meta_lr)
```
?

**Comment by brando90** [comment]
note: also `ways` is not being used in `fast_adapt`

**Comment by arijitthegame** [comment]
@brando90 lightning does support multi-gpu support generally. But how lightning + l2l +multi-gpus work is something I have not figured out.

**Comment by arijitthegame** [comment]
@brando90 I played with cherry a bit but using schedulers with cherry should not be an issue.

**Comment by arijitthegame** [comment]
@brando90 finally for init_mthod you can do `'tcp://localhost:10001'`

See here: https://github.com/facebookresearch/simsiam

**Comment by brando90** [comment]
is there a difference between:
```
        for p in maml.parameters():
            p.grad.data.mul_(1.0 / meta_batch_size)
```
and
```
        for p in model.parameters():
            p.grad.data.mul_(1.0 / meta_batch_size)
```

?

Btw, thanks for the responses arijtthegame!

**Comment by brando90** [comment]
why do we need:
```
        # Average the accumulated gradients and optimize
        for p in maml.parameters():
            p.grad.data.mul_(1.0 / meta_batch_size)
        opt.step()  # averages gradients across all workers
```

**Comment by brando90** [comment]
is the seeding really need?
```
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
```
especially cu the seed is a function of the rank. I am assuming that is there so that when we call `tasksets.<SPLIT>.sample()` so that each rank does not just run the exact same process...right?

**Comment by brando90** [comment]
btw, I noticed that if I try to setup my own master addr and port it deadlocks...do you know why?

e.g.
```
        MASTER_ADDR = '127.0.0.1'
        MASTER_PORT = master_port
        # set up the master's ip address so this child process can coordinate
        os.environ['MASTER_ADDR'] = MASTER_ADDR
        print(f"---> {MASTER_ADDR=}")
        os.environ['MASTER_PORT'] = MASTER_PORT
        print(f"---> {MASTER_PORT=}")
       dist.init_process_group(backend, init_method=init_method, rank=rank, world_size=world_size)
```

**Comment by arijitthegame** [comment]
I had the same issues before. I just set up my training just like the SimSiam code I posted above. 

Regarding: model/maml parameters, I think you need to do maml parameters and not model. If you look at the source code for maml https://github.com/learnables/learn2learn/blob/master/learn2learn/algorithms/maml.py#L171. So you create a clone of the model and the gradient hooks so you need to update the clone and not the original model. 

I have always provided seeds for training to replicate my results as much as possible. I generally have a function that sets all the possible seeds. Some of them are unnecessary if you are doing set numpy seed and not using np. 

Finally regarding division by meta_batch_size, it is a common practice if you accumulate gradients, you divide by the number of gradient steps for example here: https://discuss.pytorch.org/t/how-to-implement-accumulated-gradient/3822/8 
Since batch is constant, you can either scale the loss or scale the params. 

Hope this answers your questions.

**Comment by brando90** [comment]
@arijitthegame Thanks for the responses!

Another question. I wanted to run another instance of my maml distributed training but I get that because the other session is running there is a TCP issue...but the code I pasted deadlocked...so how do I fix this so that it doesn't deadlock?

See error:
```
(meta_learning_a100) [miranda9@hal-dgx diversity-for-predictive-success-of-meta-learning]$ bash main_dgx.sh 
-- Start my submission file
SLURM_JOBID = 26987
CUDA_VISIBLE_DEVICES
1,7,6,5,4,3,2,0
torch.cuda.device_count is:
8
---- Running your python main ----
pid = 57050
CUDA_VISIBLE_DEVICES = 1,7,6,5,4,3,2,0
SLURM_JOBID = 26987
-- Done submitting job in dgx A100-SXM4-40G
(meta_learning_a100) [miranda9@hal-dgx diversity-for-predictive-success-of-meta-learning]$ WARNING:__main__:*****************************************
Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed. 
*****************************************
ERROR:torch.distributed.elastic.multiprocessing.errors.error_handler:{
  "message": {
    "message": "RuntimeError: Address already in use",
    "extraInfo": {
      "py_callstack": [
        "  File \"/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/runpy.py\", line 197, in _run_module_as_main\n    return _run_code(code, main_globals, None,\n",
        "  File \"/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/runpy.py\", line 87, in _run_code\n    exec(code, run_globals)\n",
        "  File \"/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/run.py\", line 702, in <module>\n    main()\n",
        "  File \"/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py\", line 370, in wrapper\n    error_handler.record_exception(e)\n",
        "  File \"/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/elastic/multiprocessing/errors/error_handler.py\", line 86, in record_exception\n    _write_error(e, self._get_error_file_path())\n",
        "  File \"/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/elastic/multiprocessing/errors/error_handler.py\", line 26, in _write_error\n    \"py_callstack\": traceback.format_stack(),\n"
      ],
      "timestamp": "1644026084"
    }
  }
}
Traceback (most recent call last):
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/run.py", line 702, in <module>
    main()
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 361, in wrapper
    return f(*args, **kwargs)
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/run.py", line 698, in main
    run(args)
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/run.py", line 689, in run
    elastic_launch(
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/launcher/api.py", line 116, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/launcher/api.py", line 237, in launch_agent
    result = agent.run()
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/elastic/metrics/api.py", line 125, in wrapper
    result = f(*args, **kwargs)
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/elastic/agent/server/api.py", line 700, in run
    result = self._invoke_run(role)
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/elastic/agent/server/api.py", line 822, in _invoke_run
    self._initialize_workers(self._worker_group)
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/elastic/metrics/api.py", line 125, in wrapper
    result = f(*args, **kwargs)
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/elastic/agent/server/api.py", line 670, in _initialize_workers
    self._rendezvous(worker_group)
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/elastic/metrics/api.py", line 125, in wrapper
    result = f(*args, **kwargs)
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/elastic/agent/server/api.py", line 530, in _rendezvous
    store, group_rank, group_world_size = spec.rdzv_handler.next_rendezvous()
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/elastic/rendezvous/static_tcp_rendezvous.py", line 55, in next_rendezvous
    self._store = TCPStore(
RuntimeError: Address already in use
```

**Comment by arijitthegame** [comment]
@brando90  Does either of the suggestion help? Seems like there are some processes that are still occupying the port. https://discuss.pytorch.org/t/multiprocessing-failed-with-torch-distributed-launch-module/33056
https://github.com/PyTorchLightning/pytorch-lightning/issues/2081

**Comment by brando90** [comment]
> https://github.com/facebookresearch/simsiam

I tried:
```
python -m torch.distributed.run --nproc_per_node=8 ~/diversity-for-predictive-success-of-meta-learning/div_src/diversity_src/experiment_mains/main_dist_maml_l2l.py --manual_loads_name l2l_resnet12rfs_cifarfs_adam_cl_80k --init_method 'tcp://localhost:10001'
```

but that deadlocks. 

What is the other suggestion? Sorry I guess I missed it. :(

**Comment by brando90** [comment]
ah, it's not liking my attempt:
```
python -m torch.distributed.run --nproc_per_node=8 --master_addr="127.0.0.1'" --master_port=$RANDOM ~/diversity-for-predictive-success-of-meta-learning/div_src/diversity_src/experiment_mains/main_dist_maml_l2l.py --manual_loads_name l2l_resnet12rfs_cifarfs_adam_cl_80k > $OUT_FILE &
```
but get this error:
```
ValueError: The hostname of the rendezvous endpoint '127.0.0.1':10764' 
must be a dot-separated list of labels, an IPv4 address, or an IPv6 address.
```

**Comment by brando90** [comment]
ah darn it there is an extra `'`

**Comment by brando90** [comment]
@seba-1511 sorry for the direct ping seba, but do you mind sharing what your `init_method` is? 

Whenever I try things I can't seem to run the file you gave twice even if I try to give it different address, ports etc. I want to run more than 2 experiments at once if possible... :)

**Comment by brando90** [comment]
crossposted this here for now: https://stackoverflow.com/questions/70994896/how-to-run-torch-distributed-run-such-that-one-can-run-two-instances-of-the-scri

**Comment by brando90** [comment]
weird now I am getting:
```
(meta_learning_a100) [miranda9@hal-dgx diversity-for-predictive-success-of-meta-learning]$ python -m torch.distributed.run --nproc_per_node=8 --master_addr="127.0.0.1" --master_port=$RANDOM ~/diversity-for-predictive-success-of-meta-learning/div_src/diversity_src/experiment_mains/main_dist_maml_l2l.py --manual_loads_name l2l_resnet12rfs_cifarfs_adam_cl_80k
WARNING:__main__:*****************************************
Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed. 
*****************************************
ERROR:torch.distributed.elastic.multiprocessing.errors.error_handler:{
  "message": {
    "message": "RuntimeError: Permission denied",
    "extraInfo": {
      "py_callstack": [
        "  File \"/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/runpy.py\", line 197, in _run_module_as_main\n    return _run_code(code, main_globals, None,\n",
        "  File \"/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/runpy.py\", line 87, in _run_code\n    exec(code, run_globals)\n",
        "  File \"/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/run.py\", line 702, in <module>\n    main()\n",
        "  File \"/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py\", line 370, in wrapper\n    error_handler.record_exception(e)\n",
        "  File \"/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/elastic/multiprocessing/errors/error_handler.py\", line 86, in record_exception\n    _write_error(e, self._get_error_file_path())\n",
        "  File \"/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/elastic/multiprocessing/errors/error_handler.py\", line 26, in _write_error\n    \"py_callstack\": traceback.format_stack(),\n"
      ],
      "timestamp": "1644029376"
    }
  }
}
Traceback (most recent call last):
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/run.py", line 702, in <module>
    main()
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 361, in wrapper
    return f(*args, **kwargs)
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/run.py", line 698, in main
    run(args)
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/run.py", line 689, in run
    elastic_launch(
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/launcher/api.py", line 116, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/launcher/api.py", line 237, in launch_agent
    result = agent.run()
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/elastic/metrics/api.py", line 125, in wrapper
    result = f(*args, **kwargs)
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/elastic/agent/server/api.py", line 700, in run
    result = self._invoke_run(role)
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/elastic/agent/server/api.py", line 822, in _invoke_run
    self._initialize_workers(self._worker_group)
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/elastic/metrics/api.py", line 125, in wrapper
    result = f(*args, **kwargs)
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/elastic/agent/server/api.py", line 670, in _initialize_workers
    self._rendezvous(worker_group)
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/elastic/metrics/api.py", line 125, in wrapper
    result = f(*args, **kwargs)
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/elastic/agent/server/api.py", line 530, in _rendezvous
    store, group_rank, group_world_size = spec.rdzv_handler.next_rendezvous()
  File "/home/miranda9/miniconda3/envs/meta_learning_a100/lib/python3.9/site-packages/torch/distributed/elastic/rendezvous/static_tcp_rendezvous.py", line 55, in next_rendezvous
    self._store = TCPStore(
RuntimeError: Permission denied
```

**Comment by arijitthegame** [comment]
It seemed it ran but ran out of memory.

**Comment by arijitthegame** [comment]
Or this for the permission issue https://discuss.pytorch.org/t/tcpstore-permission-denied/57762

**Comment by brando90** [comment]
> It seemed it ran but ran out of memory.

yea I noticed. But I see that the jobs already running only occupy `5035MiB` of memory and it should have `40537MiB`. Perhaps pytorch does something else and dissalows me to allocate it...?
```
(meta_learning_a100) [miranda9@hal-dgx diversity-for-predictive-success-of-meta-learning]$ nvidia-smi
Fri Feb  4 20:54:03 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 450.156.00   Driver Version: 450.156.00   CUDA Version: 11.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  A100-SXM4-40GB      On   | 00000000:07:00.0 Off |                    0 |
| N/A   36C    P0   131W / 400W |   7233MiB / 40537MiB |     74%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   1  A100-SXM4-40GB      On   | 00000000:0F:00.0 Off |                    0 |
| N/A   35C    P0   174W / 400W |   4892MiB / 40537MiB |     83%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   2  A100-SXM4-40GB      On   | 00000000:47:00.0 Off |                    0 |
| N/A   40C    P0   175W / 400W |   4892MiB / 40537MiB |     77%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   3  A100-SXM4-40GB      On   | 00000000:4E:00.0 Off |                    0 |
| N/A   33C    P0    79W / 400W |   5038MiB / 40537MiB |     77%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   4  A100-SXM4-40GB      On   | 00000000:87:00.0 Off |                    0 |
| N/A   42C    P0    91W / 400W |   5038MiB / 40537MiB |     81%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   5  A100-SXM4-40GB      On   | 00000000:90:00.0 Off |                    0 |
| N/A   42C    P0    88W / 400W |   5038MiB / 40537MiB |     64%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   6  A100-SXM4-40GB      On   | 00000000:B7:00.0 Off |                    0 |
| N/A   47C    P0   173W / 400W |   5038MiB / 40537MiB |     74%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   7  A100-SXM4-40GB      On   | 00000000:BD:00.0 Off |                    0 |
| N/A   41C    P0   128W / 400W |   5038MiB / 40537MiB |     78%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A     17552      C   ...3/envs/recon3d/bin/python     2195MiB |
|    0   N/A  N/A    188942      C   ..._learning_a100/bin/python     5035MiB |
|    1   N/A  N/A    188928      C   ..._learning_a100/bin/python     4889MiB |
|    2   N/A  N/A    188941      C   ..._learning_a100/bin/python     4889MiB |
|    3   N/A  N/A    188939      C   ..._learning_a100/bin/python     5035MiB |
|    4   N/A  N/A    188938      C   ..._learning_a100/bin/python     5035MiB |
|    5   N/A  N/A    188931      C   ..._learning_a100/bin/python     5035MiB |
|    6   N/A  N/A    188930      C   ..._learning_a100/bin/python     5035MiB |
|    7   N/A  N/A    188929      C   ..._learning_a100/bin/python     5035MiB |
+-----------------------------------------------------------------------------+
```

**Comment by brando90** [comment]
> Or this for the permission issue https://discuss.pytorch.org/t/tcpstore-permission-denied/57762

perhaps it chose a random port that is a bad one. Unlucky. That is why I had my python script find a valid usable one rather than random numbers...

**Comment by arijitthegame** [comment]
Yes torch allocates some memory for the process. Yeah that's what I think. DDP for me is always full of challenges for me everyday.

**Comment by brando90** [comment]
> Yes torch allocates some memory for the process. Yeah that's what I think. DDP for me is always full of challenges for me everyday.

Oh, it seems that even if I run seba's script with the right args (with my model) I get a memory issue. Ok, I suppose I can go to sleep knowing I did it right :) 

But it's puzzling why it allocates 5 gigs then it has 40 spare and doesn't let me run another one.

Perhaps it's because we are not using `torchrun`?

**Comment by brando90** [comment]
> Yes torch allocates some memory for the process. Yeah that's what I think. DDP for me is always full of challenges for me everyday.

but why does it allocate GPU memory? I am puzzled....

**Comment by arijitthegame** [comment]
Here: https://stackoverflow.com/questions/58149401/why-pytorch-needs-much-more-memory-than-it-should

Two suggestions regarding the memory issue: Try FoMAML as opposed to MAML and turn off validation. You can also try reducing the number of shots too.

**Comment by arijitthegame** [comment]
I use DDP everyday and I do not use torch run.

**Comment by brando90** [comment]
@arijitthegame thnx for the suggestions! Will think about what to do...still puzzled and sad it crashes the way it does when it seems to only allocate 5 gigs for my expts...

**Comment by arijitthegame** [comment]
That is weird. Something might be hogging the memory. Pytorch only allocates a small amount not this much. Maybe add some memory profiler in various parts of your script and see what is causing oom.

**Comment by brando90** [comment]
> That is weird. Something might be hogging the memory. Pytorch only allocates a small amount not this much. Maybe add some memory profiler in various parts of your script and see what is causing oom.

hmmm... will think about how to go about this. What puzzles me is that the rest of the GPU is free according to nvidia-smi so would my python profiler show that the two don't agree?

**Comment by brando90** [comment]
@arijitthegame out of curiosity, do you know why the `maml.clone()` is needed? Would it break l2l's MAML implementation?

**Comment by seba-1511** [comment]
> @seba-1511 sorry for the direct ping seba, but do you mind sharing what your init_method is?

I only run 1 meta-learning job at a time using DDP. As @arijitthegame pointed out, you should be able to get multiple distributed runs going by changing the address. If you're using a single machine (or distributed filesystem) I've had success with different init files (ie, `.dist_init1`, `.dist_init2`, etc).

Hope this works for you.

**Comment by arijitthegame** [comment]
@brando90 See this discussion about the clone function https://github.com/learnables/learn2learn/issues/169

**Comment by brando90** [comment]
> > @seba-1511 sorry for the direct ping seba, but do you mind sharing what your init_method is?
> 
> I only run 1 meta-learning job at a time using DDP. As @arijitthegame pointed out, you should be able to get multiple distributed runs going by changing the address. If you're using a single machine (or distributed filesystem) I've had success with different init files (ie, `.dist_init1`, `.dist_init2`, etc).
> 
> Hope this works for you.

Do you mind sharing 2 example init files? That way I can see the contents in it? To get an idea of what to do...thanks Seba!

**Comment by seba-1511** [comment]
Those are empty files, which are created by PyTorch to synchronize all runs. The only requirement is that they must be accessible to all processes.

**Comment by brando90** [comment]
@seba-1511 this code is not needed if running serially:
```
        for p in meta_learner.parameters():
            p.grad.data.mul_(1.0 / meta_batch_size)
        outer_opt.step()  # averages gradients across all workers
```
and can be replaced by:
```
        if args.world_size > 1:
            for p in meta_learner.parameters():
                p.grad.data.mul_(1.0 / meta_batch_size)
        outer_opt.step()  # averages gradients across all workers
```

right?

**Comment by brando90** [comment]
Actually, why is that loop needed in the parallel case?

**Comment by brando90** [comment]
@seba-1511 

Feel free to ignore this if it's not urgent.

Would the following work:

1. wrap your taskset in a collate function of a standard pytorch dataloader.
2. then in the collate, sample multiple times according to the batch size.
3. then use DDP with the normal pytorch data loader (no need for cherry I think).

Would this work?

**Comment by arijitthegame** [comment]
@brando90  I believe this will work. I almost to do it but not quite. We split the tasks across machines/nodes but you make sure that you do not split the data for a given task across the machines. A pseudo code to do that will be : 

- Create dataloader and distributed dataparallel for task 
- For the trainer for the task, you can not use another dataloader as that would cause errors. Maybe there is a way to sort that out but I could not figure out so I wrote a (slow) data shuffling and batching using just basic python (and some pytorch).

This will get you started on what you want to do.

**Comment by brando90** [comment]
I know without more code for context it's hard but -- for the distributed MAML I am getting the code deadlocks. Its odd. What are possible sources of deadlocking in learn2learn? Can fetching the data cause this? @seba-1511
